### PR TITLE
Add conformance tests with requestHeaderModifier

### LINKED ISSUE #2037: Conformance: HTTPRoute Filter Combinations
All HTTPRoute filters now have conformance tests as covered by https://github.com/kubernetes-sigs/gateway-api/issues/1103. Now we need some tests to cover combinations of them. At minimum we should aim to have all filters tested in combination with one other filter. Here's an example of what that could look like:

**Core Combinations**
* [ ] Request Header Modifier and Request Redirect

**Core and Extended Combinations**
* [ ] Request Header Modifier and URL Rewrite
* [ ] Request Redirect and Request Mirror

**Extended Combinations**
* [ ] Response Header Modifier and Request Mirror

### PR DESCRIPTION
**Pull request by Treenhan**
**What type of PR is this?**
/kind test
/area conformance

**What this PR does / why we need it**:
Adding conformance testing by combining one Core Capability (i.e. requestHeaderModifier) wither other Extended Capabilities (i.e. requestMirror, responseHeaderModifier, and urlRewrite)

**Which issue(s) this PR fixes**:
Fixes #2037

```release-note
NONE
```

### CODE REVIEW

File: conformance/tests/httproute-redirect-path.yaml [hunk 1]
```diff
@@ -70,4 +70,4 @@ spec:
         path:
           type: ReplaceFullPath
           replaceFullPath: /replacement-full
-        statusCode: 301
+        statusCode: 301
```
Comment (LiorLieberman) [inline conformance/tests/httproute-redirect-path.yaml:73]: nit: newline

File: conformance/tests/httproute-response-header-modifier.go [hunk 2]
```diff
@@ -168,6 +168,43 @@ var HTTPRouteResponseHeaderModifier = suite.ConformanceTest{
...
+				Headers: map[string]string{
+					"X-Header-Remove": "val",
+				},
+			},
+			BackendSetResponseHeaders: map[string]string{
+				"X-Header-Set-2":    "set-val-2",
+				"X-Header-Add-2":    "add-val-2",
+				"X-Header-Remove-2": "remove-val-2",
+				"Another-Header":    "another-header-val",
+				"X-Header-Remove-1": "val",
```
Comment (LiorLieberman) [inline conformance/tests/httproute-response-header-modifier.go:183]: for consistency, maybe `remove-val-1` ?

File: conformance/tests/httproute-response-header-modifier.go [hunk 3]
```diff
@@ -168,6 +168,43 @@ var HTTPRouteResponseHeaderModifier = suite.ConformanceTest{
...
+				},
+				AbsentHeaders: []string{"X-Header-Remove"},
+			},
+			Response: http.Response{
+				Headers: map[string]string{
+					"X-Header-Set-1": "header-set-1",
+					"X-Header-Set-2": "header-set-2",
+					"X-Header-Add-1": "header-add-1",
+					"X-Header-Add-2": "add-val-2,header-add-2",
+					"X-Header-Add-3": "header-add-3",
```
Comment (LiorLieberman) [inline conformance/tests/httproute-response-header-modifier.go:201]: seems like `X-Header-Add-3` is checking the same functionality as  `X-Header-Add-1` so we can remove one of them (both here and from the yaml)

**Review by LiorLieberman — COMMENTED** [review]
Thanks @Treenhan ! 

Aside from a few small comments looks good. 

/ok-to-test

**Comment by Xunzhuo** [comment]
Plz sign CLA first @Treenhan, Thanks!

**Comment by Treenhan** [comment]
> Plz sign CLA first @Treenhan, Thanks!

I just created CLA approval request, currently waiting on that to go through

File: conformance/tests/httproute-redirect-path.yaml (hunk 1 above)
Comment (Treenhan) [reply conformance/tests/httproute-redirect-path.yaml]: Done

File: conformance/tests/httproute-response-header-modifier.go (hunk 2 above)
Comment (Treenhan) [reply conformance/tests/httproute-response-header-modifier.go]: Done

File: conformance/tests/httproute-response-header-modifier.go (hunk 3 above)
Comment (Treenhan) [reply conformance/tests/httproute-response-header-modifier.go]: Done

File: conformance/tests/httproute-rewrite-host.go [hunk 4]
```diff
@@ -73,6 +73,27 @@ var HTTPRouteRewriteHost = suite.ConformanceTest{
...
+					Headers: map[string]string{
+						"X-Header-Remove": "val",
+					},
+				},
+				ExpectedRequest: &http.ExpectedRequest{
+					Request: http.Request{
+						Path: "/rewrite-host-and-modify-headers",
+						Host: "test.example.org",
+						Headers: map[string]string{
+							"X-Header-Add": "add-appends-values",
```
[the remark this answers is not in the document]
Comment (Treenhan) [reply conformance/tests/httproute-rewrite-host.go]: Done

File: conformance/tests/httproute-rewrite-path.yaml [hunk 5]
```diff
@@ -33,3 +33,26 @@ spec:
...
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /full/rewrite-path-and-modify-headers
+    filters:
+    - type: URLRewrite
+      urlRewrite:
+        path:
+          type: ReplaceFullPath
+          replaceFullPath: /test
```
[the remark this answers is not in the document]
Comment (Treenhan) [reply conformance/tests/httproute-rewrite-path.yaml:45]: Done

File: conformance/tests/httproute-response-header-modifier.go [hunk 6]
```diff
@@ -168,6 +168,43 @@ var HTTPRouteResponseHeaderModifier = suite.ConformanceTest{
...
+				"X-Header-Add-2":    "add-val-2",
+				"X-Header-Remove-2": "remove-val-2",
+				"Another-Header":    "another-header-val",
+				"X-Header-Remove-1": "val",
+			},
+			ExpectedRequest: &http.ExpectedRequest{
+				Request: http.Request{
+					Path: "/response-and-request-header-modifiers",
+					Headers: map[string]string{
+						"X-Header-Add": "add-appends-values",
```
[the remark this answers is not in the document]
Comment (Treenhan) [reply conformance/tests/httproute-response-header-modifier.go]: Done

File: conformance/tests/httproute-rewrite-host.yaml [hunk 7]
```diff
@@ -27,3 +27,24 @@ spec:
...
+            - name: X-Header-Set
+              value: set-overwrites-values
+            add:
+              - name: X-Header-Add
+                value: add-appends-values
+            remove:
+              - X-Header-Remove
+      backendRefs:
+        - name: infra-backend-v2
+          port: 8080
```
[the remark this answers is not in the document]
Comment (Treenhan) [reply conformance/tests/httproute-rewrite-host.yaml]: Done

**Comment by LiorLieberman** [comment]
Thanks @Treenhan 
Looks good to me! I ran the tests and everything works as expected. 

/lgtm
@robscott can you approve ?

**Comment by Treenhan** [comment]
Hi @robscott could you please take a look?

File: conformance/tests/httproute-request-mirror.go [hunk 8]
```diff
@@ -58,6 +58,28 @@ var HTTPRouteRequestMirror = suite.ConformanceTest{
...
+					Request: http.Request{
+						Path: "/mirror-and-modify-headers",
+						Headers: map[string]string{
+							"X-Header-Add":        "header-val-1",
+							"X-Header-Add-Append": "append-val-1,header-val-2",
+							"X-Header-Set":        "set-overwrites-values",
+						},
+					},
+					AbsentHeaders: []string{"X-Header-Remove"},
+				},
```
Comment (robscott) [inline conformance/tests/httproute-request-mirror.go:79]: As far as I can tell, this doesn't actually verify that the request to the mirrored backend received the modified headers (https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/http/mirror.go#L32). Assuming my understanding is correct, we should probably either add a comment here or create an issue to track improving that. @LiorLieberman would probably know best.

File: conformance/tests/httproute-response-header-modifier.go [hunk 9]
```diff
@@ -168,6 +168,44 @@ var HTTPRouteResponseHeaderModifier = suite.ConformanceTest{
...
+					"X-Header-Remove":     "remove-val",
+					"X-Header-Add-Append": "append-val-1",
+				},
+			},
+			BackendSetResponseHeaders: map[string]string{
+				"X-Header-Set-2":    "set-val-2",
+				"X-Header-Add-2":    "add-val-2",
+				"X-Header-Remove-2": "remove-val-2",
+				"Another-Header":    "another-header-val",
+				"X-Header-Remove-1": "remove-val-1",
```
Comment (robscott) [inline conformance/tests/httproute-response-header-modifier.go:184]: Maybe add one here that is not modified in any way to show that it passes all the way through.

**Review by robscott — COMMENTED** [review]
Thanks @Treenhan! This is great! Nothing blocking here, but a couple small nits. Would appreciate a review from others here as well:

/cc @arkodg @mlavacca @michaelbeaumont @sunjayBhatia @Xunzhuo

File: conformance/tests/httproute-rewrite-path.go [hunk 10]
```diff
@@ -70,6 +70,48 @@ var HTTPRouteRewritePath = suite.ConformanceTest{
...
+						"X-Header-Add-Append": "append-val-1",
+					},
+				},
+				ExpectedRequest: &http.ExpectedRequest{
+					Request: http.Request{
+						Path: "/test",
+						Headers: map[string]string{
+							"X-Header-Add":        "header-val-1",
+							"X-Header-Add-Append": "append-val-1,header-val-2",
+							"X-Header-Set":        "set-overwrites-values",
```
Comment (sunjayBhatia) [inline conformance/tests/httproute-rewrite-path.go:87]: Anywhere we're asserting a set/overwrite header modification, we should maybe have that header in the original request (e.g. here add `X-Header-Set` to the original request)

**Review by sunjayBhatia — COMMENTED** [review]
test all pass with contour 👍🏽

File: conformance/tests/httproute-rewrite-path.go (hunk 10 above)
Comment (Treenhan) [reply conformance/tests/httproute-rewrite-path.go:87]: I have added `X-Header-Set` as suggested, thanks Sunjay!

File: conformance/tests/httproute-response-header-modifier.go (hunk 9 above)
Comment (Treenhan) [reply conformance/tests/httproute-response-header-modifier.go:184]: Thanks for the suggestion Rob! I have added it in.

File: conformance/tests/httproute-request-mirror.go (hunk 8 above)
Comment (Treenhan) [reply conformance/tests/httproute-request-mirror.go:79]: Thanks both for pointing this out, I created a separate issue to improve this later https://github.com/kubernetes-sigs/gateway-api/issues/2142

**Comment by robscott** [comment]
Thanks @Treenhan!

/lgtm
/approve

**Pull request merged**